src/corell/src/lib.rs

```diff
@@ -0,0 +1,136 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#[macro_use]
+extern crate log;
+
+use std::fmt::Debug;
+use std::hash::Hash;
+use std::any::Any;
+
+pub mod command;
+pub mod factory;
+pub mod format;
+pub mod memory;
+
+/// Draw vertex count.
+pub type VertexCount = u32;
+/// Draw number of instances
+pub type InstanceCount = u32;
+/// Number of vertices in a patch
+pub type PatchSize = u8;
+
+/// Describes what geometric primitives are created from vertex data.
+#[derive(Copy, Clone, Debug, Hash, PartialEq, Eq)]
+#[repr(u8)]
+pub enum Primitive {
+    /// Each vertex represents a single point.
+    PointList,
+    /// Each pair of vertices represent a single line segment. For example, with `[a, b, c, d,
+    /// e]`, `a` and `b` form a line, `c` and `d` form a line, and `e` is discarded.
+    LineList,
+    /// Every two consecutive vertices represent a single line segment. Visually forms a "path" of
+    /// lines, as they are all connected. For example, with `[a, b, c]`, `a` and `b` form a line
+    /// line, and `b` and `c` form a line.
+    LineStrip,
+    /// Each triplet of vertices represent a single triangle. For example, with `[a, b, c, d, e]`,
+    /// `a`, `b`, and `c` form a triangle, `d` and `e` are discarded.
+    TriangleList,
+    /// Every three consecutive vertices represent a single triangle. For example, with `[a, b, c,
+    /// d]`, `a`, `b`, and `c` form a triangle, and `b`, `c`, and `d` form a triangle.
+    TriangleStrip,
+    /// Patch list,
+    /// used with shaders capable of producing primitives on their own (tessellation)
+    PatchList(PatchSize),
+}
+
+/// A type of each index value in the slice's index buffer
+#[derive(Eq, Ord, PartialEq, PartialOrd, Hash, Copy, Clone, Debug)]
+#[allow(missing_docs)]
+#[repr(u8)]
+pub enum IndexType {
+    U16,
+    U32,
+}
+
+pub trait Instance {
+    type B: Backend;
+
+    fn create() -> Self;
+
+    // TODO: Use an iterator instead of Vec?
+    fn enumerate_physical_devices(&self) -> Vec<<<Self as Instance>::B as Backend>::PhysicalDevice>;
+}
+
+pub trait PhysicalDevice {
+    type B: Backend;
+
+    fn open(&self) -> (<<Self as PhysicalDevice>::B as Backend>::Device, Vec<<<Self as PhysicalDevice>::B as Backend>::CommandQueue>);
+    fn get_info(&self) -> PhysicalDeviceInfo;
+}
+
+#[derive(Clone, Debug)]
+pub struct PhysicalDeviceInfo {
+    /// Phyiscal device name
+    pub name: String,
+    /// Vendor PCI id of the physical device
+    pub vendor: usize,
+    /// PCI id of the physical device
+    pub device: usize,
+    /// The device is based on a software rasterizer
+    pub software_rendering: bool,
+}
+
+pub trait Device {
+    
+}
+
+pub trait CommandQueue {
+    type B: Backend;
```

Comment:
This wouldn't cause any issues.
I noticed only that it might only become a bit more verbose (e.g https://github.com/msiglreith/gfx/blob/gfx_next/src/core_next/src/lib.rs#L101)